keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,141 @@
+
+## Table of Contents
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Challenges](#challenges)
+    * [Root Object](#root-object)
+    * [Child Objects](#child-objects)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks](#drawbacks)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+Link to tracking issue: [kubernetes/kubectl#572](https://github.com/kubernetes/kubectl/issues/572)
+
+## Summary
+
+`kubectl apply --prune` has been in Alpha for more than a year. This design proposes a better approach of handling `--prune` and move it to Beta. This feature targets 1.15 release.
+
+## Motivation
+
+`kubectl apply` operates an aggregation of Kubernetes Objects. While `kubectl apply` is widely used in different workflows, users expect it to be able to sync the objects on the cluster with configuration saved in files, directories or repos. That means, when `kubectl apply` is run on those configurations, the corresponding objects are expected to be correctly created, updated and pruned. 
+
+  Currently `kubectl apply --prune` is still in Alpha. It is based on label matching. Users need to be really careful to use this command, so that they don't delete objects unintentionally. There is a list of [issues]([kubernetes/kubectl#572]) observed by users. The most noticeable issue happens when changing directories or working with directories with the same labels. When two directories share the same label. `kubectl apply --prune` from one directory will delete objects from the other directory.
+
+ In this design, we take an overview of the pruning challenges and propose a better pruning solution. This new approach is able to work with or to be utilized by generic configuration management tools.
+
+### Goals
+
+Propose a solution for better `kubectl apply --prune` so that
+- it can find the correct set of objects to delete
+- it can avoid mess up when changing directories
+
+### Non-Goals
+
+- Make `--prune` default on
+- Improvement on Garbage Collections
+
+## Proposal
+
+### Challenges
+The challenges of pruning are mainly from two aspects.
+- Find out correct objects to delete. When a directory is applied after some updates, it loses the track of objects that are previously applied from the same directory.
+- Delete objects at proper time. Assume the set of objects to be deleted is available, when to delete them is another challenge. Those objects could be still needed during rolling out other objects.
+
+In this design, we focus on a solution for the first challenge. For the second challenge, we can blacklist certain types such as ConfigMaps and Secrets for deletion. Generally deleting objects at proper time can be better addressed by Garbage Collection.
```

Comment:
There are plenty of other extension types now like configmap and secret.  How do they get blacklisted?

Reply:
iiuc the objects that will participate in the prune must also be white listed by having them select the root object and by including them in the root object by GVKN? Does this proposal allow users to create objects (e.g. ConfigMaps and Secrets) that will not participate in pruning? If so are you really saying that you will (in addition to requiring white listing) allow for the implementation of explicit black listing in the future?

Reply:
We can also add the blacklist types into `data` to allow customizing the blacklist.
```
data:
  object1: ...
  object2:...
  object3:...
  blacklist1: namespace1:group1:version1:kind1
  blacklist2: namespace2:group2:version2:kind2
```

Reply:
Instead of blacklisting, would it work to only perform deletion after rollout has been completed if we could figure out how to detect that?   I can imagine other resources being dependencies that shouldn't be deleted until a rollout is complete.

@smarterclayton 

> There are plenty of other extension types now like configmap and secret. How do they get blacklisted?

I believe the issue with configmaps and secrets that is being addressed is due to 2 factors - 1) some tools generate them based on the contents of another file - so they are likely to be replaced by a new instance without the user explicitly deleting them and 2) the old instances will be referenced by Pods that may not have been rolled out during a rolling update.  Pruning them would then delete them before they stopped being used by these Pods.

Are you thinking that this issue or other issues might apply to extension or other types?

Reply:
nm, I think we could do this without whitelisting.

Reply:
If someone has multiple applications (think of it as a term for grouping) in a namespace, will a prune of application A cause the assets for application B to be deleted? Unless they know about each other and have them blacklisted from pruning?

Reply:
Ah, nevermind... I see the diff discussed elsewhere. Ignore the prev comment.